### ISSUE TITLE
How to set correct blockly for Linux system?

### ISSUE THREAD

**Issue by WolfOptics** [BIPES/BIPES#79]
Hi sirs,
I got some problem using blockly in linux system as shown as picture.
https://drive.google.com/file/d/1n4EhUH57qYAP2y8ppRB7OT9e7SsM2ooh/view?usp=sharing

Could you help me? It was NOT workable in my raspberry pi 3B+.

Best Regards,

**Comment by JorgeGMarques** [comment]
@rafaelaroca Can you provide some insight on how to expose the python shell to BIPES?

**Comment by WolfOptics** [comment]
Hi sir,
 The information was shown as below in my Linux's terminal after I push "Run block based program" in BIPES.

[https://drive.google.com/file/d/1FrJYa0WBIC-BUK8VhASxTx0uSY1r5ynG/view?usp=sharing](url)

Best Regards,

**Comment by rafaelaroca** [comment]
Hello @WolfOptics, sorry for the delay.

Did you try to run SocketServerWebREPL on the board?

Like this:

First start SocketServerWebREPL:

$ git clone https://github.com/BIPES/SocketServerWebREPL
$ cd SocketServerWebREPL/
$ python2 socketserverWebREPL.py 
BIPES Project - Bridge started
Now, access the Python Console from a Web Browser using a WebREPL client
For example: https://micropython.org/webrepl/#127.0.0.1:8266/

You can run the commands above on your Raspberry Pi or any other Linux device. Even your PC, for testing. After SocketServerWebREPL is started, got to BIPES and fill the IP address:

![image](https://user-images.githubusercontent.com/31751413/146658690-a8666bb8-49ee-46e1-a869-cd438f26eb83.png)

Click on the "Network" text, which means "Connect":

You should see this message at the terminal:

![image](https://user-images.githubusercontent.com/31751413/146658709-c57428a0-6c96-4e14-befe-b4663c6d53f8.png)


Type s and you should get the Python REPL:

![image](https://user-images.githubusercontent.com/31751413/146658718-79d001d2-a11d-4676-a15f-4b241d7533d2.png)

Then you can try some commands:

![image](https://user-images.githubusercontent.com/31751413/146658729-64c52638-1070-4e72-9c34-5f2e8084d306.png)

Now, seems that there is a bug for sending block based code to the host. We will check it and get back.

Hope this helps for a first test.

Best regards.

**Comment by andrewg78** [comment]
I can confirm the issue
![raspberrypi2](https://user-images.githubusercontent.com/2042344/147385696-721add2e-047f-461b-8df7-38e22c08209b.png)
.

**Comment by rafaelaroca** [comment]
Hi @andrewg78, thanks for testing. Nice to know. Did the console output "ABC", or the program crashed? Thanks for helping debug BIPES!

**Comment by andrewg78** [comment]
1. I can send blocks several times
 through the 'Run block based program' It hits rasbperry PI every time, but there is a new line char after the 'b' ![printabc](https://user-images.githubusercontent.com/2042344/147393679-f7e95fe2-3f4e-4c4c-aac9-d7deeb796c83.png)
![rintab](https://user-images.githubusercontent.com/2042344/147393694-a1bcd5b3-3f2b-467e-bf16-fa62780f5d82.png)

2. When I refresh BIPES site (F5), SocketServerWebREPL receives last char from the BIPES console in the infinite loop!!!  ![infinite-loop](https://user-images.githubusercontent.com/2042344/147393682-dffefd79-4ae6-4dd5-a376-36c2207f3d7b.png)
3. Python Remote Shell sends python code properly No issue here!!!
4. After I send block through 'Run block based program'  prompt stop responding on  ' Python Remote Shell ' ( I do not see what I'm typing)

**Comment by rafaelaroca** [comment]
Hello @andrewg78, just updated https://github.com/BIPES/SocketServerWebREPL. Could you please update your SocketServerWebREPL and try again? Running programs is now reliable, but there are still other bugs. But one thing at a time :-)

**Comment by rafaelaroca** [comment]
> Hello @andrewg78, just updated https://github.com/BIPES/SocketServerWebREPL. Could you please update your SocketServerWebREPL and try again? Running programs is now reliable, but there are still other bugs. But one thing at a time :-)

By the way, here is a working example:

Link to the program: 
http://bipes.net.br/beta2/ui/#9ec4ro

Program:
![image](https://user-images.githubusercontent.com/31751413/147398423-70ddc1c0-362d-4c25-9a03-6e09696de9a2.png)



Result:
![image](https://user-images.githubusercontent.com/31751413/147398411-5ee32a17-2edd-4faf-bc95-13378359920c.png)

**Comment by andrewg78** [comment]
Hi. I see some progress, although there are still some glitches ;)
So:
1. What is the expected outcome on the Python Remote Shell and on the LInux terminal? I see two different results. Should we have the same outputs ?
2. There is a syntax Error at the end.
![syntax_error](https://user-images.githubusercontent.com/2042344/147420360-417bb5bb-a2a8-45db-9b2f-ad3a8b0bc5ea.png)

3. When Python Shell is not started(s key not pressed), but block based program is run, we got this
![python-shell-not-started](https://user-images.githubusercontent.com/2042344/147463760-b98b6e5f-0531-4b11-acad-03dfd63244df.png)

4. when I start socket server as a non sudo user, program is executed from blocks but we got a lot of 'IndentationError: unexpected indent' 
![non-sudo](https://user-images.githubusercontent.com/2042344/147464450-687efedb-2443-4e68-8208-b895864eb2e1.png)

5. When I run with sudo (sudo python socketserverWebREPL.py) there is NO issue with 'unexpected indent'
6. When I use mouse scroll in the Remote Python Shell, BIPES disconnects from the  socket server and we  must again click on Connect to restore the connection
![scroll](https://user-images.githubusercontent.com/2042344/147464808-a5575a4f-97ec-4c5f-a98a-a61e81cdf41d.png)
7. System commands output always '0' 
![system_commands](https://user-images.githubusercontent.com/2042344/147465418-39efa6af-7803-4c4b-8dc1-4ec9884da24d.png)

**Comment by rafaelaroca** [comment]
> Hi. I see some progress, although there are still some glitches ;) So:
> 
> 1. What is the expected outcome on the Python Remote Shell and on the LInux terminal? I see two different results. Should we have the same outputs ?
> 2. There is a syntax Error at the end.
>    ![syntax_error](https://user-images.githubusercontent.com/2042344/147420360-417bb5bb-a2a8-45db-9b2f-ad3a8b0bc5ea.png)
> 3. When Python Shell is not started(s key not pressed), but block based program is run, we got this
>    ![python-shell-not-started](https://user-images.githubusercontent.com/2042344/147463760-b98b6e5f-0531-4b11-acad-03dfd63244df.png)
> 4. when I start socket server as a non sudo user, program is executed from blocks but we got a lot of 'IndentationError: unexpected indent'
>    ![non-sudo](https://user-images.githubusercontent.com/2042344/147464450-687efedb-2443-4e68-8208-b895864eb2e1.png)
> 5. When I run with sudo (sudo python socketserverWebREPL.py) there is NO issue with 'unexpected indent'
> 6. When I use mouse scroll in the Remote Python Shell, BIPES disconnects from the  socket server and we  must again click on Connect to restore the connection
>    ![scroll](https://user-images.githubusercontent.com/2042344/147464808-a5575a4f-97ec-4c5f-a98a-a61e81cdf41d.png)
> 7. System commands output always '0'
>    ![system_commands](https://user-images.githubusercontent.com/2042344/147465418-39efa6af-7803-4c4b-8dc1-4ec9884da24d.png)

Hi @andrewg78, we are advancing. Thank you for helping!

1. Run Linux Command is expected to run a command on the remote board, but without getting the output. I use it, for example, to call mpg123 and start playing mp3 music, or to call speak so that the remote device can speak text from a BIPES command. So, the outputs are different, because the command output is shown on the remote terminal, but we could change it so that it is also shown on the BIPES console.

2. Sorry, this is a bug. Tested and confirmed. We will fix it, but not that that code execution works. This error may happen due to some garbage at the end of the buffer.

3. This is ok. Without Python shell started, no Python statement should work. No commands should be sent before starting Python session (s sent).

4 and 5. This issue did not occur to me for both normal execution and execution with sudo. Maybe your default python interpreter is different between these users (user default user and root)?

6. Will check it later. I did not test/replicate the issue.

7. Ok, the command specified on os.system() runs, as you showed on the remote terminal. What happens is that the STDOUT and STDERR of the executed command is not redirected to BIPES console. Actually, we did not implement this, so we can have it on our roadmap. Anyway, if you need the program output you can build a block for that!! :-)
with the command: x = os.popen('date').read()

For example:

![image](https://user-images.githubusercontent.com/31751413/147494756-37fdcbed-77ea-4e29-be65-e616f1b66dc3.png)

The above example creates a function that receives an argument and execute the argument as a Linux command, getting the result and returning to the main program to be used.

**Comment by rafaelaroca** [comment]
Sorry, did not test the last example. This one is tested / working:

![image](https://user-images.githubusercontent.com/31751413/147495185-b6530cc7-f764-4b03-8f04-b293b57b7194.png)

"Run Linux Command and get output" was created using the "Functions" item of the toolbox. The function receives a command as argument, builds the command with os.popen(command).read() and evaluates the result to the out variable, which is returned. Note the test with several commands.

Direct link to the program:
http://bipes.net.br/beta2/ui/#2azfub

**Comment by rafaelaroca** [comment]
> Hi. I see some progress, although there are still some glitches ;) So:
> 
> 1. What is the expected outcome on the Python Remote Shell and on the LInux terminal? I see two different results. Should we have the same outputs ?
> 2. There is a syntax Error at the end.
>    ![syntax_error](https://user-images.githubusercontent.com/2042344/147420360-417bb5bb-a2a8-45db-9b2f-ad3a8b0bc5ea.png)
> 3. When Python Shell is not started(s key not pressed), but block based program is run, we got this
>    ![python-shell-not-started](https://user-images.githubusercontent.com/2042344/147463760-b98b6e5f-0531-4b11-acad-03dfd63244df.png)
> 4. when I start socket server as a non sudo user, program is executed from blocks but we got a lot of 'IndentationError: unexpected indent'
>    ![non-sudo](https://user-images.githubusercontent.com/2042344/147464450-687efedb-2443-4e68-8208-b895864eb2e1.png)
> 5. When I run with sudo (sudo python socketserverWebREPL.py) there is NO issue with 'unexpected indent'
> 6. When I use mouse scroll in the Remote Python Shell, BIPES disconnects from the  socket server and we  must again click on Connect to restore the connection
>    ![scroll](https://user-images.githubusercontent.com/2042344/147464808-a5575a4f-97ec-4c5f-a98a-a61e81cdf41d.png)
> 7. System commands output always '0'
>    ![system_commands](https://user-images.githubusercontent.com/2042344/147465418-39efa6af-7803-4c4b-8dc1-4ec9884da24d.png)

Hello. Item 2 (syntax error at the end) is now fixed. Please, get the latest version on github.

![image](https://user-images.githubusercontent.com/31751413/147563454-ca7db674-071c-41ed-a078-ed82ce912ce8.png)

**Comment by andrewg78** [comment]
Hi. Confirmed:
issue number 2 is fixed! Good job!
issues 4 and 5 also fixed ( sudo not needed on the newest version)
Regarding the issue no3 .I would argue here ;)
This is not user friendly. For instance, in order to get the file list I must first go to Console Tab  and Enable Remote Shell with the 's' key.
Can we have Python Shell started automatically after the connection is established ?

**Comment by andrewg78** [comment]
Issue no 6 - I updated description in the original post. This one should be closed as well.

**Comment by rafaelaroca** [comment]
Hi @andrewg78 !

Agree with your point on issue number 3. It is not friendly, however, a minimal authentication would be nice. In that way, I have just updated https://github.com/BIPES/SocketServerWebREPL so that SocketServerWebREPL asks for a password when the connection is established. By default, the password is "bipes". You can have the password typed on the connection dialog box, so it will be automatically sent, and the connection will be established directly, as you suggested. 

Password should be typed here (and will be saved):
![image](https://user-images.githubusercontent.com/31751413/147591439-cbe6d592-9f9c-4436-afd7-e5b42fa1fbf2.png)

Well, finally, now the shell is started automatically, as you suggested!

**Comment by andrewg78** [comment]
Issue no 3 is resolved. Gr8!!!
I think we should move some of your responses and code examples to WIKI or to documentation page. Do you have one?  I can copy-paste it to wiki right away. Just let me know.
There is still minor issue with websocket server when BIPES page is refreshed, but I will create separate issue on that. This thread can be closed I think

**Comment by rafaelaroca** [comment]
Hi @andrewg78. Great! Closing this issue! Thank you very much for helping debug it!
